### ISSUE TITLE
Best way to handle CKErrorLimitExceeded in CloudKit modification operations

### ISSUE THREAD

**Issue by swiftlyfalling** [danthorpe/Operations#288]
According to [Apple's docs](https://developer.apple.com/library/ios/documentation/CloudKit/Reference/CKModifyRecordsOperation_class/index.html#//apple_ref/occ/cl/CKModifyRecordsOperation):

> To ensure the speed of fetching and saving records, the server may reject large operations. When this occurs, a block reports the CKErrorLimitExceeded error. **Your app should handle this error, and refactor the operation into multiple smaller batches.**

(Emphasis mine.)

Such refactoring is complicated by CloudKit features like atomic commits, and is thus very application-specific.

This applies to at least _CKModifyRecordsOperation_ (but could theoretically apply to other operations that take multiple items for input, such as the other _CKModify*_ operations, _CKMarkNotificationsReadOperation_, etc). 

I noticed that you have **BatchedCloudKitOperation** for fetch operations (_CKFetchNotificationChangesOperation_ / _CKFetchRecordChangesOperation_), which is fantastic. But what would be the best way to handle the other side of this - for batched modifications - when the server returns an error of CKErrorLimitExceeded and the client needs to handle refactoring the operation into smaller batches? (Presumably including preserving order.)

**Comment by danthorpe** [comment]
Hi @swiftlyfalling - thanks for the excellent feedback & question.

First off, the `BatchedCloudKitOperation` class only supports usage with `CKFetchNotificationChangesOperation` and `CKFetchRecordChangesOperation`, as these are the only two `CKOperation` classes which have support for receiving `moreComing` and `resultsList`. So... I guess it's for when the client asks for potentially lots of data, and it will receive it in batches. I think I could definitely make the names clearer for this. For `CKModifyRecordsOperation`, where the client is sending potentially large amounts of data to the server, I guess this class won't be relevant currently.

However, we should still be able to handle this error type. It just might be a bit tricky, as CloudKit doesn't give you any indication of what a smaller batch size should be. And it might be a bit fiddly too. So, lets look at `CKModifyRecordsOperation` as an example. Our basic strategy will be to:

1. Add an error handler for this code, inside the error handler, we'll capture (`unowned`) the operation.
2. Bisect the data being sent, into left and right halves. The left is the first half, and we want to send it first.
3. Create a new operation, configured as the current operation. Set the left half of data on it.
4. Implement the configure block (which will be applied when new attempts are made) for the right half.
  a. In this block, apply the suggested configuration
  b. Set the right half of the data
  c. Add the left half operation as a dependency.
5. Add the left half operation as a child operation of the current operation (recall that `CloudKitOperation` inherits from `GroupOperation` so it has its own queue, and you can add operations to it.
6. Return the error response (the new configure block)

```swift
let operation = CloudKitOperation { CKModifyRecordsOperation() } // etc
// etc further configuration of the operation such as container etc

operation.setErrorHandlerForCode(.LimitExceeded) { [unowned operation] (error, log, suggested) -> (delay: Delay?, configure: OPRCKOperation<CKModifyRecordsOperation> -> Void)? in
            
    // What we need to do is retry the operation but with "smaller batches"
    // But how do we handle this?
            
    // First, we can log that we've got an error
    log.warning("Received CloudKit error Limit Exceeded: \(error)")
            
    // We should return an ErrorResponse?, which is a tuple, with an
    // Delay? and a configuration block. The configuration block is applied
    // to a new instance of the CKModifyRecordsOperation.
            
    // The strategy will be to half the recordsToSave (and any other
    // properties being sent). We'll send the first half in a new
    // CloudKitOperation, which we'll make a dependency of the
    // current operation which we will modify & retry.
            
    // Create a new operation to bisect the data
    let left = CloudKitOperation { CKModifyRecordsOperation() }
            
    // Update config
    left.container = operation.container
    left.database = operation.database // etc
            
    // Set the data to the first half of the data
    // This could probably be better!
    let rightRecordsToSave: [CKRecord]? = .None
    if let recordsToSave = operation.recordsToSave {
        left.recordsToSave = Array(recordsToSave.prefixUpTo(recordsToSave.count/2))
        rightRecordsToSave = Array(recordsToSave.suffixFrom(recordsToSave.count/2))
    }
            
    let configure = { toConfigure in
                
        // Apply the suggested configuration (will include the database, container etc)
        suggested.configure(toConfigure)
                
        // Set the properties on the retried operation, to the right half
        toConfigure.recordsToSave = rightRecordsToSave
                
        // Add the left half as a dependency of the right half,
        // this is to ensure a consistent ordering.
        toConfigure.addDependency(left)
    }
            
    // Add the left half as a child of the original operation
    operation.addOperation(left)
            
    // Return the error response
    return (suggested.delay, configure)
}
```
 
This strategy is certainly a bit fiddly. And it will probably require a bit of fiddling to get it working correctly. One thing I can see immediately, is that we need to set the error handlers on the new operation of the left half.

But, what is quite nice about this, is that we add as a child a new `CloudKitOperation` and then depend on that to succeed before doing the left half. And this divide and conquer approach should work multiple times assuming that the error handler can be re-applied.

Will obviously need looking at more once I've got v2.10 done.

**Comment by swiftlyfalling** [comment]
@danthorpe , thank you for the suggestions (and the framework). Am fiddling around with your suggestion.

One other concern with the way things are currently structured is how to properly handle CKErrorPartialFailure. I opened a new issue for that: #290

**Comment by danthorpe** [comment]
Okay great. Yeah, come back at me with ideas too, and we can improve this. I'm not currently using the CloudKit stuff in my day-to-day work, so I'm a little bit off the boil on some of the nitty-gritty. But, I'm really pleased with the abstractions here, so, I'm eager to fix any practical issues or limitations with it :)

**Comment by swiftlyfalling** [comment]
Is there any easy way that I'm missing to apply the suggested.configure block to the new ("left") operation? That would remove a whole bunch of code simply setting left.bla = operation.bla.

**Comment by swiftlyfalling** [comment]
Also, just to confirm Apple's recommendation for adjusting the batch size in response to CKErrorLimitExceeded:

> The general guideline we give developers to handle this error is simply divide the number of items in your batch by half and issue two operations instead of one.
> 
> And recursively do that if those operations encounter the same error again.

Source: `WWDC 2015: CloudKit Tips and Tricks (Session 715)` ([Video](https://developer.apple.com/videos/play/wwdc2015/715/)) ([Transcript](http://asciiwwdc.com/2015/sessions/715))

**Comment by danthorpe** [comment]
Hey @swiftlyfalling - sorry - I didn't understand your comment, 2nd previous comment here?

**Comment by danthorpe** [comment]
Oh @swiftlyfalling I see, sorry - yes, you can do this:

```swift
left.addConfigureBlock { suggested.configure($0) }
```

This is a public API on `RepeatedOperation` which lets you add (FIFO style) configuration block which then get applied to the target operation before they are executed.

**Comment by swiftlyfalling** [comment]
@danthorpe, that addConfigureBlock API is exactly what I was looking for. Perfect!

In regards to supporting CKErrorLimitExceeded, I'm not sure it should be handled by default by `CloudKitOperation`. 

While Apple's recommendation is to split the recordsToSave and recordIDsToDelete in half, there is no guarantee that this can be done safely by the framework with a simple bisection. There are cases where a simple bisection can directly result in failures to save records.

For example: 

> An app could submit a `CKModifyRecordsOperation` that contains a set of `recordsToSave` that include `CKReference`s (references to other `CKRecords`). This will work, regardless of the order of the `CKRecord`s in the `recordsToSave array`, as long as all the `CKReference`s in each `CKRecord` are either: 
(a) already stored on the server
(b) exist in the recordsToSave array (i.e. in the same transaction).
> 
> If you do a simple bisection, you run the risk of splitting the record pointed to by a CKRecord's CKReference into a later operation (since the original order in the recordsToSave array is irrelevant as they are all in the same original transaction), which will cause saving the CKRecords *containing* those CKReferences to fail.

That said, the simple bisection method can be useful in *some* cases, so I think it makes sense to add - just maybe as a separate new **BatchedCloudKitModifyOperation**(?) class (with documentation about the limits involving transactions with records containing CKReferences).

**Comment by danthorpe** [comment]
@swiftlyfalling yeah, the way that I would do it, is similar to how the current error handler works - it requires an optional response, but it provides a suggested response.

So, the suggested response would be the result of the bisection (i.e. it would contain the left and right array of the batched properties to save). The framework consumer can implement this, and inspect the suggestion, and make adjustments as needed. All you'd need to do is return the left & the right - and the framework will take that and deal with the dependencies and other configuration stuff.

**Comment by swiftlyfalling** [comment]
Thinking about this more, there is potentially a less-simple bisection algorithm that helps handle the first CKReference-related failure case above.

It would involve ensuring a sort of topological sorting of recordsToSave where:

- All CKRecords that are referenced by a CKReference come _before_ the item that references them.
- Unless there is a circular reference (which CloudKit explicitly allows), in which case all applicable CKRecords *must* be in the same CKModifyRecordsOperation.
- Or unless the CKRecord pointed to by a CKReference does not exist in the recordsToSave array, in which case it is safe to ignore. (As it is either already on the server, being handled by a prior Operation, or will fail _just like the original operation would have_)

Or, in other words, CKReferences are considered a back-reference - a directed edge from vertex _u_ (**target of CKReference**) -> vertex _v_ (**the CKRecord containing the CKReference**). Assuming vertex _u_ is present in the recordsToSave, _u_ comes before _v_ in the ordering.

So basically, all records except those involved in circular references loops / directed cycles can be ordered as above and bisected. All records that _are_ involved in circular reference loops / directed cycles need to be in the same operation.

Apple explicitly allows circular references:

> It is permissible to create circular owning references for a set of records.

Source: [CKReference Class Reference](https://developer.apple.com/library/ios/documentation/CloudKit/Reference/CKReference_class/)

So nothing prevents you from having:
CKRecordA["b"] -> CKRecordB
CKRecordB["c"] -> CKRecordC
CKRecordC["a"] -> CKRecordA
or:
CKRecordA["b"] -> CKRecordB
CKRecordB["c"] -> CKRecordC
CKRecordB["a"] -> CKRecordA
CKRecordC["a"] -> CKRecordA

Of course, theoretically, there still could be a set of CKRecords in a directed cycle that results in CKErrorLimitExceeded. So if the framework's bisection attempt fails to split the recordsToSave at all, it would need to stop trying, and give the framework consumer an opportunity to handle the error.

**Comment by swiftlyfalling** [comment]
@danthorpe, re: [your suggestion above](https://github.com/danthorpe/Operations/issues/288#issuecomment-212932549)

Are you saying that if the framework consumer does _not_ implement this new handler, that CloudKitOperation would _not_ attempt bisection?

(To clarify: I don't think the framework should attempt any sort of manipulation of the input to a CKModify operation by default. Only if explicitly opted-into by the framework consumer.)

**Comment by swiftlyfalling** [comment]
Also, just like the CKErrorPartialFailure case, this could presumably apply (at a minimum) to any CKOperations that take a set of input for modification:

the `recordsToSave` and `recordIDsToDelete` for CKModifyRecordsOperation
the `recordZonesToSave` and `recordZoneIDsToDelete` for CKModifyRecordZonesOperation
the `subscriptionsToSave` and `subscriptionIDsToDelete` for CKModifySubscriptionsOperation
the `notificationIDs` for CKMarkNotificationsReadOperation

i.e. All specified input parameters for the operations above may need to be bisected in response to CKErrorLimitExceeded (or otherwise handled).

**Comment by danthorpe** [comment]
Yes - exactly, it would be an error handler which does most of the heavy lifting for you. You still need to opt-in, and you need to confirm the bisect data, but it would basically do everything else for you.

**Comment by danthorpe** [comment]
@swiftlyfalling - sorry, only just read your previous comment about ensuring that the bisecting is "valid" - makes a lot of sense. This is something that could be looked at after the infrastructure for providing the suggested bisecting is done. So, to start with, I will look at just doing the simple option. And refine it later.

**Comment by bunkersmith** [comment]
@danthorpe I stumbled on your code snippet above posted just over four years ago and used it to solve a CKModifyRecordsOperation .LimitExceeded problem of my own. Many thanks!!!
